### PR TITLE
[Frontend] Add tool_choice=required support for GPT-OSS Harmony models

### PR DESCRIPTION
**Pull request by gkswns0531**
## Summary

Add `tool_choice="required"` support for GPT-OSS Harmony models via EBNF grammar, and fix the Harmony render path to actually apply tool parser grammar constraints.

## Problem

GPT-OSS models use the Harmony chat format with channel-based output (`analysis`, `commentary`, `final`). Two issues prevented `tool_choice="required"` from working:

1. **No grammar enforcement mechanism** — The standard JSON schema approach (base class `adjust_request`) constrains output to a JSON array, which is incompatible with Harmony's channel-based format.
2. **Harmony render path bypasses `adjust_request()`** — `render_chat()` routes GPT-OSS requests through `_make_request_with_harmony()` instead of `_preprocess_chat()`, so `tool_parser.adjust_request()` was never called. Even with a correct grammar implementation, it was never applied.

Without this fix, tool call generation depends entirely on the model's natural behavior (~85% success rate at default temperature).

### EBNF Grammar (commit 1)

Use an EBNF grammar compiled by xgrammar to constrain token-level generation while preserving the Harmony channel structure:

- **Allows** analysis blocks (chain-of-thought reasoning)
- **Allows** commentary preambles (visible text before tool calls)
- **Requires** at least one tool call (`commentary to=functions.X`)
- **Blocks** the final channel entirely (not defined in grammar)

```
root       ::= non_tool_block* tool_block more_tool*
non_tool_block ::= ("analysis" | "commentary") "<|message|>" content "<|end|>" "<|start|>" "assistant" "<|channel|>"
tool_block ::= "commentary to=" func_name "<|message|>" content "<|end|>" "<|call|>"
more_tool  ::= "<|start|>" "assistant" "<|channel|>" non_tool_block* tool_block
func_name  ::= "functions.get_weather" | "functions.calculate" | ...
content    ::= ([^<] | "<" [^|])*
```

### Harmony Render Path Fix (commit 2)

Add `adjust_request()` call in the Harmony branch of `render_chat()` so grammar constraints are actually applied for GPT-OSS models.

### Why EBNF instead of LogitsProcessor?

The [previous implementation](https://github.com/vllm-project/vllm/pull/33306/commits/0ee72e8ae93fa001d892d57f170586bb8d1f8b59) used a custom `PatternForcedSequenceLogitsProcessor`. This was changed because:

1. **LogitsProcessor refactoring** — `LogitsProcessor` is being refactored for model runner v2 ([comment](https://github.com/vllm-project/vllm/pull/33306#discussion_r2876521083))
2. **Reasoning preservation** — EBNF grammar allows analysis/commentary channels, preserving chain-of-thought ([concern](https://github.com/vllm-project/vllm/pull/33306#issuecomment-2648889953))
3. **Smaller diff** — Leverages existing `structured_outputs` + xgrammar path, no model runner internals modified

## Changes

- **`vllm/entrypoints/serve/render/serving.py`**: Call `tool_parser.adjust_request()` in the Harmony render path (`_make_request_with_harmony` branch) so grammar constraints are applied for GPT-OSS models.
- **`vllm/tool_parsers/openai_tool_parser.py`**: Override `adjust_request()` with EBNF grammar builder for `tool_choice=required`. Use `raw_decode()` for JSON extraction to handle trailing structural tokens.
- **`vllm/entrypoints/openai/parser/harmony_utils.py`**: Add error handling in `parse_output_into_messages` for grammar-constrained token sequences.
- **`tests/tool_parsers/test_openai_tool_parser_ebnf.py`**: Unit tests for EBNF grammar and xgrammar validation.

## Test Plan

- [x] Unit tests: xgrammar acceptance/blocking/termination tests pass
- [x] Live server batch test: **100/100 requests** produce tool calls on `gpt-oss-120b` (default temperature)
- [x] Before fix: ~85% success rate (grammar was never applied due to render path bug)
- [x] After fix: **100% success rate**

Related Issue: #33966

### CODE REVIEW

**Comment by gkswns0531** [comment]
Hello, @chaunceyjiang 
Just a friendly reminder about this PR when you have a moment. I'd really appreciate it if you could take a look whenever your schedule allows.
Happy to address any feedback or questions. 
Thank you!

**Comment by gkswns0531** [comment]
Hi @chaunceyjiang , just a gentle follow-up on this PR. I understand you must be busy, so no rush at all.
If this issue has already been resolved in the meantime, I'll go ahead and close this PR. Otherwise, if there's a preferred alternative approach or any concerns about the implementation, I'd be more than happy to hear your thoughts and make adjustments accordingly.
Thanks again for your time!

**Comment by gkswns0531** [comment]
@jonoillar  Thanks for the thorough review and great questions!

1. Multiple analysis rounds:

You're right that the Harmony format supports multiple `<|channel|>analysis` messages and preambles (plain `commentary` without `to=`) in a single response. The current implementation intentionally trades off these for guaranteed tool call generation. I also explored forcing only `commentary` to allow the model to generate preambles before tool calls. However, the model would often generate a commentary message and then immediately emit `<|return|>`, ending the response without actually making a tool call. So forcing `commentary to=` was necessary to ensure the model commits to the tool call path.

In production, many systems depend on tool calls happening reliably at specific points — orchestration agents, pipelines with mandatory function calls, etc. Most models support `tool_choice="required"` out of the box, but GPT-OSS Harmony models currently lack this enforcement, which was the core motivation for this PR.

`tool_choice="auto"` is where the model should have full freedom — multiple analysis rounds, preambles, choosing whether to call tools or respond directly, etc. But `tool_choice="required"` explicitly signals that the caller needs a tool call to happen, no matter what. In that context, I believe it's important to have an option that guarantees 100% tool call generation, even if it means sacrificing some reasoning depth. Users who want the model's full reasoning capability can always use `tool_choice="auto"`.


2. The bad_words approach:

This was actually my first implementation. I registered the `final` token in `bad_words` to prevent the model from entering the final channel. This doesn't block multiple analysis rounds, and it improved the tool call success rate from ~80% to ~97%, but the model could produce variant tokens like `" final"` (with a leading space) or `"finally"` that bypassed the filter, introducing 2-3% errors.

The current LogitsProcessor approach solves this with positive enforcement — instead of trying to block all possible variants of unwanted tokens, it forces the exact required sequence (`commentary to=`) at the channel decision point, guaranteeing 100% tool call generation.

3. Schema enforcement after forcing:

Once the model enters the `commentary to=` path, GPT-OSS models are trained to reliably generate well-formed tool calls — selecting the correct function name, producing valid JSON arguments, and terminating cleanly with `<|call|>`. This isn't theoretically enforced at the logits level like xgrammar, but in practice, across hundreds of test runs, the model has consistently produced valid tool calls once it's on the correct channel path.

---

Thanks again for taking the time to review this so carefully. I'm happy to adjust the approach if the community has a different preference — always opento working together toward what benefits the project most.

File: vllm/v1/sample/logits_processor/pattern_forced_sequence.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    """Per-request state for pattern-triggered forced sequence."""
+
+    state: ForcingState
+    forcing_pos: int
+    output_ids: list[int]
+    trigger_pattern: list[int]
+    forced_sequence: list[int]
+
+
+class PatternForcedSequenceLogitsProcessor(LogitsProcessor):
```
Comment (chaunceyjiang) [inline vllm/v1/sample/logits_processor/pattern_forced_sequence.py]: Oh, by the way, `LogitsProcessor` is currently undergoing refactoring, since model runner v2 will restructure this part of the code.

This PR might need to wait until model runner v2 is officially enabled by default before we continue moving it forward.

Comment (gkswns0531) [reply vllm/v1/sample/logits_processor/pattern_forced_sequence.py]: Thanks! Understood — I'll watch for the model runner v2 progress and update this PR accordingly!

**Review by sfeng33 — COMMENTED** [review]
I was wondering if you have considered solving the problem using an alternative approach: structural tag, e.g. define tags for analysis and commentary to=functions.*

That way the reasoning is preserved, no change is needed on the model runner. Structure tag has been supported in xgrammar/guidance.

**Comment by gkswns0531** [comment]
@sfeng33 @bbrowning Here are the BFCL multi-turn results:

**gpt-oss-20b** (H200 SXM - 1 GPU)

| Category | Base auto | PR auto | Base required | PR required |
|---|---|---|---|---|
| base | 30.50% | 29.00% | 30.00% | 6.00% |
| miss_func | 26.50% | 26.00% | 20.00% | 4.00% |
| miss_param | 30.00% | 28.00% | 29.00% | 3.00% |
| long_context | 4.00% | 5.00% | 4.50% | 0.00% |
| **Overall** | **22.75%** | **22.00%** | **20.88%** | **3.25%** |

**gpt-oss-120b** (H200 SXM - 8 GPU, PR only) 

| Category | PR auto | PR required |
|---|---|---|
| base | 46.00% | 29.50% |
| miss_func | 35.00% | 14.00% |
| miss_param | 38.50% | 16.50% |
| long_context | 7.50% | 5.00% |
| **Overall** | **31.75%** | **16.25%** |

- **No regression in auto mode** — Base 22.75% vs PR 22.00%, within noise.
- **Base required ≈ Base auto** (20.88% vs 22.75%) — confirms `tool_choice=required` had no effect before this PR, i.e., no grammar enforcement was happening.
- **PR required scores drop in multi-turn** — the grammar now forces tool calls on every turn, including turns where the model should respond with text. Confirms the implementation is working as intended.
